### PR TITLE
Fixes #17978 - custom facts from commandline works

### PR DESCRIPTION
**Pull request by lzap**
They were ignored.

### CODE REVIEW

**Comment by lzap** [comment]
@iNecas here is your ISO: http://downloads.theforeman.org/discovery/scratch-lzap-custfact/

File: root/usr/lib64/ruby/vendor_ruby/discovery/screen/countdown.rb [hunk 1]
```diff
@@ -23,7 +22,7 @@ def screen_countdown discovery_only = false
   unless discovery_only
     sec = secs
     while sec > 0
-      l_press.set_text(press_key_text)
+      l_press.set_text("< " + _('Press any key') + " ... (#{sec}s) >")
```
Comment (lzap) [inline root/usr/lib64/ruby/vendor_ruby/discovery/screen/countdown.rb]: "Press any key" message countdown was not working I noticed.

File: root/usr/lib64/ruby/vendor_ruby/discovery/screen/facts.rb [hunk 2]
```diff
@@ -46,7 +41,7 @@ def screen_facts mac, proxy_url, proxy_type
     if perform_upload(proxy_url, proxy_type, custom_facts)
       [:screen_status, generate_info(' - ' + _('awaiting kexec into installer'))]
     else
-      :screen_welcome
+      [:screen_status, generate_info(' - ' + _('fact upload failed'))]
```
Comment (lzap) [inline root/usr/lib64/ruby/vendor_ruby/discovery/screen/facts.rb]: This is also an improvement - when PXE-less upload fails, you are taken to status page which is better than welcome page (to investigate logs or resend and stuff).

**Comment by iNecas** [comment]
I can confirm from testing that the custom facts upload works now. Few notices:

* the status doesn't show the "waiting for kexec" message - not fatal, but nice to have
* when trying to kexec, it started downloading, but ended up with undefined method join for string:

```
Jan 10 21:47:44 fdi foreman-proxy[1080]: Downloading: http://example.com/pxeboot/vmlinuz
Jan 10 21:47:44 fdi foreman-proxy[1080]: Starting task: /usr/bin/wget --timeout=10 --tries=3 --no-check-certificate -nv -c "http://example.com/pxeboot/vmlinuz" -O "/tmp/vmlinuz"
Jan 10 21:48:09 fdi foreman-proxy[1080]: [2329] 2017-01-10 21:48:09 URL:http://example.com/pxeboot/vmlinuz [5154912/5154912] -> "/tmp/vmlinuz" [1]
Jan 10 21:48:09 fdi foreman-proxy[1080]: Downloading: http://example.com/pxeboot/initrd.img
Jan 10 21:48:09 fdi foreman-proxy[1080]: Starting task: /usr/bin/wget --timeout=10 --tries=3 --no-check-certificate -nv -c "http://example.com/pxeboot/initrd.img" -O "/tmp/initrd.img"
Jan 10 21:48:44 fdi foreman-proxy[1080]: accept: 10.0.0.1:36784
Jan 10 21:48:44 fdi foreman-proxy[1080]: Rack::Handler::WEBrick is invoked.
Jan 10 21:48:44 fdi foreman-proxy[1080]: Downloading: http://example.com/pxeboot/vmlinuz
Jan 10 21:48:44 fdi foreman-proxy[1080]: Starting task: /usr/bin/wget --timeout=10 --tries=3 --no-check-certificate -nv -c "http://example.com/pxeboot/vmlinuz" -O "/tmp/vmlinuz"
Jan 10 21:48:44 fdi foreman-proxy[1080]: Downloading: http://example.com/pxeboot/initrd.img
Jan 10 21:48:44 fdi foreman-proxy[1080]: undefined method `join' for #<String:0x007f9ab4281f08>
Jan 10 21:48:44 fdi foreman-proxy[1080]: close: 10.0.0.1:36784
Jan 10 21:47:44 fdi foreman-proxy[1080]: Downloading: http://example.com/pxeboot/vmlinuz
Jan 10 21:47:44 fdi foreman-proxy[1080]: Starting task: /usr/bin/wget --timeout=10 --tries=3 --no-check-certificate -nv -c "http://example.com/pxeboot/vmlinuz" -O "/tmp/vmlinuz"
Jan 10 21:48:09 fdi foreman-proxy[1080]: [2329] 2017-01-10 21:48:09 URL:http://example.com/pxeboot/vmlinuz [5154912/5154912] -> "/tmp/vmlinuz" [1]
Jan 10 21:48:09 fdi foreman-proxy[1080]: Downloading: http://example.com/pxeboot/initrd.img
Jan 10 21:48:09 fdi foreman-proxy[1080]: Starting task: /usr/bin/wget --timeout=10 --tries=3 --no-check-certificate -nv -c "http://example.com/pxeboot/initrd.img" -O "/tmp/initrd.img"
Jan 10 21:48:44 fdi foreman-proxy[1080]: accept: 10.0.0.1:36784
Jan 10 21:48:44 fdi foreman-proxy[1080]: Rack::Handler::WEBrick is invoked.
Jan 10 21:48:44 fdi foreman-proxy[1080]: Downloading: http://example.com/pxeboot/vmlinuz
Jan 10 21:48:44 fdi foreman-proxy[1080]: Starting task: /usr/bin/wget --timeout=10 --tries=3 --no-check-certificate -nv -c "http://example.com/pxeboot/vmlinuz" -O "/tmp/vmlinuz"
Jan 10 21:48:44 fdi foreman-proxy[1080]: Downloading: http://example.com/pxeboot/initrd.img
Jan 10 21:48:44 fdi foreman-proxy[1080]: undefined method `join' for #<String:0x007f9ab4281f08>
Jan 10 21:48:44 fdi foreman-proxy[1080]: close: 10.0.0.1:36784
```

My kexec template looked like this:

```
{
  "kernel": "boot/RedHat-7.3-x86_64-vmlinuz",
  "initram": "boot/RedHat-7.3-x86_64-initrd.img",
  "append": "ks=http://example.com/unattended/provision?token=2bacf30a-1873-4125-82a1-d82ff2c5203f&static=yes inst.ks.sendmac ip=::::::none nameserver= ksdevice=bootif BOOTIF=00-00-1a-4a-23-15-13 "
}
```

Not sure it's relevant, but it was used against sat 6.2

**Comment by lzap** [comment]
Ivan, the waiting message needs to stay, I can't tell if you are going to kexec or reboot. It depends on if you have kexec template associated or not.

The error message is weird indeed, but it's consequence of incorrect URL. You are looking for absolute HTTP addresses, these are not right. Image simply passes them to wget and something wrong happened (can't see why it's this weird message).

**Comment by lzap** [comment]
We are tracking a ticket to add a code that will call HTTP HEAD on both addresses in discovery plugin prior kexecing, so user will be informed early about this.

**Comment by lzap** [comment]
For the record, the error seems to be from Sinatra: http://projects.theforeman.org/issues/18092

**Comment by lzap** [comment]
@iNecas anything else?

**Comment by iNecas** [comment]
Nothing else

**Comment by lzap** [comment]
Thanks @timogoebel if you find time to merge this one as well.

**Pull request merged**